crates/daphne-worker/src/aggregator/queues/async_aggregator.rs

```diff
@@ -0,0 +1,342 @@
+}
+
+impl CapnprotoPayloadEncode for AsyncAggregationMessage<'_> {
+    type Builder<'a> = queue_messages_capnp::async_aggregation_message::Builder<'a>;
+
+    fn encode_to_builder(&self, mut builder: Self::Builder<'_>) {
+        let Self {
+            version,
+            part_batch_sel,
+            agg_job_id,
+            initialize_reports,
+            taskprov_advertisement,
+        } = self;
+        builder.set_version((*version).into());
+        part_batch_sel.encode_to_builder(builder.reborrow().init_partial_batch_selector());
+        agg_job_id.encode_to_builder(builder.reborrow().init_aggregation_job_id());
+        initialize_reports.encode_to_builder(builder.reborrow().init_initialize_reports());
+        match taskprov_advertisement.as_deref() {
+            Some(ta) => builder
+                .init_taskprov_advertisement()
+                .set_some(ta.into())
+                .unwrap(),
+            None => builder.init_taskprov_advertisement().set_none(()),
+        }
+    }
+}
+
+impl CapnprotoPayloadDecode for AsyncAggregationMessage<'static> {
+    type Reader<'a> = queue_messages_capnp::async_aggregation_message::Reader<'a>;
+
+    fn decode_from_reader(reader: Self::Reader<'_>) -> capnp::Result<Self>
+    where
+        Self: Sized,
+    {
+        Ok(Self {
+            version: reader.get_version()?.into(),
+            agg_job_id: <_>::decode_from_reader(reader.get_aggregation_job_id()?)?,
+            part_batch_sel: <_>::decode_from_reader(reader.get_partial_batch_selector()?)?,
+            initialize_reports: <_>::decode_from_reader(reader.get_initialize_reports()?)?,
+            taskprov_advertisement: match reader.get_taskprov_advertisement()?.which()? {
+                queue_messages_capnp::option::Which::None(()) => None,
+                queue_messages_capnp::option::Which::Some(text) => Some(text?.to_string()?),
+            },
+        })
+    }
+}
+
+async fn shard_reports(
+    durable: Do<'_>,
+    task_id: &TaskId,
+    time_precision: Time,
+    agg_job_id: AggregationJobId,
+    reports: impl Iterator<Item = ReportMetadata>,
+) -> Result<HashSet<ReportId>, storage::Error> {
+    let mut shards = HashMap::<_, Vec<_>>::new();
+    for r in reports {
+        let epoch = r.time - (r.time % time_precision);
+        let shard = r.id.shard(NonZeroUsize::new(1024).unwrap());
+        shards.entry((epoch, shard)).or_default().push(r.id);
+    }
+
+    futures::stream::iter(shards)
+        .map(|((epoch, shard), report_ids)| async move {
+            durable
+                .with_retry()
+                .request(Command::Check, (task_id, epoch, shard))
+                .encode(&replay_checker::Request {
+                    report_ids: report_ids.into(),
+                    aggregation_job_id: agg_job_id,
+                })
+                .send::<replay_checker::Response>()
+                .await
+                .map(|r| r.duplicates)
+        })
+        .buffer_unordered(6)
```

Comment:
How did you pick `6`? Also, perhaps define this as a `const` so we can give it a name that reflects its purpose.